service/frontend/health_check.go

```diff
@@ -0,0 +1,150 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package frontend
+
+import (
+	"context"
+
+	"google.golang.org/grpc"
+	healthpb "google.golang.org/grpc/health/grpc_health_v1"
+
+	enumsspb "go.temporal.io/server/api/enums/v1"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/cache"
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/membership"
+	"go.temporal.io/server/common/primitives"
+	"go.temporal.io/server/service/matching"
+)
+
+type (
+	HealthCheckClient struct {
+		clients     []healthpb.HealthClient
+		serviceName string
+		failureRate dynamicconfig.FloatPropertyFn
+	}
+
+	HealthCheckClientFactory struct {
+		membershipMonitor membership.Monitor
+		rpcFactory        common.RPCFactory
+		config            *Config
+		connectionCache   cache.Cache
+	}
+
+	healthCheckResponse struct {
+		*healthpb.HealthCheckResponse
+		err error
+	}
+)
+
+func (c *HealthCheckClient) Check(ctx context.Context) enumsspb.HealthState {
+	respCh := make(chan healthCheckResponse, len(c.clients))
+	for _, client := range c.clients {
+		go func(hc healthpb.HealthClient) {
+			resp, err := hc.Check(ctx, &healthpb.HealthCheckRequest{Service: c.serviceName})
+			respCh <- healthCheckResponse{
+				HealthCheckResponse: resp,
+				err:                 err,
+			}
+		}(client)
+	}
+
+	var failureCount float64
+	for i := 0; i < len(c.clients); i++ {
+		resp := <-respCh
+		if resp.err != nil || resp.GetStatus() != healthpb.HealthCheckResponse_SERVING {
+			failureCount++
+		}
+	}
+
+	close(respCh)
+	if (failureCount / float64(len(c.clients))) < c.failureRate() {
+		return enumsspb.HEALTH_STATE_NOT_SERVING
+	}
+	return enumsspb.HEALTH_STATE_SERVING
+}
+
+func NewHealthCheckClientFactory(
+	membershipMonitor membership.Monitor,
+	rpcFactory common.RPCFactory,
+	config *Config,
+) HealthCheckClientFactory {
+	factory := HealthCheckClientFactory{
+		membershipMonitor: membershipMonitor,
+		rpcFactory:        rpcFactory,
+		config:            config,
+		connectionCache:   cache.NewLRU(500, nil),
```

Comment:
:grimacing:   why does this have its own connection cache? it should share a cache with the existing clients

Reply:
I guess that requires some refactoring in the rpcFactory impl, but yeah we should reuse the connection.

Reply:
I guess I need some refactor here as I was trying to use the grpc health check api. But I think I need to introduce deep health check to both matching and frontend.

Reply:
I think this requires more change. I will remove this for now and move to a different PR.